src/Symfony/Component/Messenger/MessageBus.php

```diff
@@ -35,17 +35,27 @@ public function __construct(iterable $middlewareHandlers = [])
         } elseif (\is_array($middlewareHandlers)) {
             $this->middlewareAggregate = new \ArrayObject($middlewareHandlers);
         } else {
-            $this->middlewareAggregate = new class() {
-                public $aggregate;
-                public $iterator;
+            // This class expects 'middlewareAggregate' property is an instance
+            // of IteratorAggregate So we wrap $middlewareHandlers in to it, and
+            // we cache the results in order to work even with generator.
+            $this->middlewareAggregate = new class($middlewareHandlers) implements \IteratorAggregate {
+                private $middlewareHandlers;
+                private $cachedIterator;
+
+                public function __construct($middlewareHandlers)
+                {
+                    $this->middlewareHandlers = $middlewareHandlers;
+                }
 
                 public function getIterator()
                 {
-                    return $this->aggregate = new \ArrayObject(iterator_to_array($this->iterator, false));
+                    if (null === $this->cachedIterator) {
+                        $this->cachedIterator = new \ArrayObject(iterator_to_array($this->middlewareHandlers, false));
```

Comment:
This is not the same: returning always the same instance won't "rewind" the iterator.
The goal of the anonymous class is to lazy load the handlers, then remove itself.

Reply:
Why do I always understand things after posted :)
So, because `ArrayObject` is itself an `IteratorAggregate`, this works, got it.

Reply:
huhu :)